File: src/csharp/Foundgine.Core/Semantic/Resolution/Lexical/SemanticLexiconProjection.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Resolution;

/// <summary>Projects a frozen semantic contract into searchable lexical documents.</summary>
public static class SemanticLexiconProjection
{
    public static IReadOnlyList<SemanticLexiconEntry> Build(SemanticContractSnapshot contract)
    {
        ArgumentNullException.ThrowIfNull(contract);

        var entries = new List<SemanticLexiconEntry>();

        foreach (var entity in contract.Entities)
        {
            entries.Add(new SemanticLexiconEntry(
                entity.Name,
                SemanticLexicalCandidateKind.Entity,
                entity.Name,
                EntityId: entity.Id,
                Aliases: entity.EffectiveAliases.Select(x => x.Name).ToArray(),
                Description: $"Semantic entity {entity.Name}."));

            entries.Add(new SemanticLexiconEntry(
                entity.Name,
                SemanticLexicalCandidateKind.Node,
                entity.Name,
                EntityId: entity.Id,
                Aliases: entity.EffectiveAliases.Select(x => x.Name).ToArray(),
                Description: $"Semantic graph node for {entity.Name}."));

            foreach (var field in entity.Fields)
            {
                entries.Add(new SemanticLexiconEntry(
                    field.Name,
                    SemanticLexicalCandidateKind.Field,
                    $"{entity.Name} {field.Name}",
                    EntityId: entity.Id,
                    FieldId: field.Id,
                    Aliases: field.EffectiveAliases.Select(x => x.Name).ToArray(),
                    Description: $"Field {entity.Name}.{field.Name}."));
            }

            foreach (var relationship in entity.Relationships)
            {
                entries.Add(new SemanticLexiconEntry(
                    relationship.Name,
                    SemanticLexicalCandidateKind.Relationship,
                    $"{entity.Name} {relationship.Name} {contract.Get(relationship.Target).Name}",
                    RelationshipId: relationship.Id,
                    SourceEntityId: entity.Id,
                    TargetEntityId: relationship.Target,
                    Aliases: relationship.EffectiveAliases.Select(x => x.Name).ToArray(),
                    Description: $"Relationship from {entity.Name} to {contract.Get(relationship.Target).Name}."));
            }
        }

        foreach (var traversal in contract.Traversals)
        {
            entries.Add(new SemanticLexiconEntry(
                traversal.Name,
                SemanticLexicalCandidateKind.Traversal,
                $"{contract.Get(traversal.Source).Name} {traversal.Name} {contract.Get(traversal.Target).Name}",
                SourceEntityId: traversal.Source,
                TargetEntityId: traversal.Target,
                Description:
                $"Logical traversal from {contract.Get(traversal.Source).Name} to {contract.Get(traversal.Target).Name}."));
        }

        return entries;
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/SemanticLexiconProjection.java
package com.foundgine.core.semantic.resolution;

import com.foundgine.core.semantic.*;
import java.util.*;

public final class SemanticLexiconProjection {
	private SemanticLexiconProjection() {
	}

	public static List<SemanticLexiconEntry> build(SemanticContractSnapshot contract) {
		Objects.requireNonNull(contract);
		var entries = new ArrayList<SemanticLexiconEntry>();
		for (var entity : contract.entities()) {
			var aliases = entity.effectiveAliases().stream().map(SemanticAlias::name).toList();
			entries.add(new SemanticLexiconEntry(entity.name(), SemanticLexicalCandidateKind.ENTITY, entity.name(),
					entity.id(), null, null, null, null, null, aliases, "Semantic entity " + entity.name() + "."));
			entries.add(new SemanticLexiconEntry(entity.name(), SemanticLexicalCandidateKind.NODE, entity.name(),
					entity.id(), null, null, null, null, null, aliases,
					"Semantic graph node for " + entity.name() + "."));
			for (var field : entity.fields()) {
				var fa = field.effectiveAliases().stream().map(SemanticAlias::name).toList();
				entries.add(new SemanticLexiconEntry(field.name(), SemanticLexicalCandidateKind.FIELD,
						entity.name() + " " + field.name(), entity.id(), null, field.id(), null, null, null, fa,
						"Field " + entity.name() + "." + field.name() + "."));
			}
			for (var relationship : entity.relationships()) {
				var target = contract.get(relationship.target());
				var ra = relationship.effectiveAliases().stream().map(SemanticAlias::name).toList();
				entries.add(new SemanticLexiconEntry(relationship.name(), SemanticLexicalCandidateKind.RELATIONSHIP,
						entity.name() + " " + relationship.name() + " " + target.name(), null, relationship.id(), null,
						entity.id(), relationship.target(), null, ra,
						"Relationship from " + entity.name() + " to " + target.name() + "."));
			}
		}
		for (var traversal : contract.traversals()) {
			var source = contract.get(traversal.source());
			var target = contract.get(traversal.target());
			entries.add(new SemanticLexiconEntry(traversal.name(), SemanticLexicalCandidateKind.TRAVERSAL,
					source.name() + " " + traversal.name() + " " + target.name(), null, null, null, traversal.source(),
					traversal.target(), null, List.of(),
					"Logical traversal from " + source.name() + " to " + target.name() + "."));
		}
		return List.copyOf(entries);
	}
}
